src/test/java/com/redhat/exhort/integration/backend/BackendUtilsTest.java

```diff
@@ -0,0 +1,162 @@
+            .filter(
+                t -> // character is hexadecimal digit a-f
+                (t.shortValue() >= 97 && t.shortValue() <= 102)
+                        // or char is one of 0-9 digits.
+                        || (t.shortValue() >= 48 && (t.shortValue() <= 57)))
+            .collect(Collectors.toList())
+            .size()
+        == requestId.length();
+  }
+
+  @Test
+  void checkGeneratedWithNoRHDATokenAndShouldBeDifferentFromEachOther() {
+
+    String reqId1 = this.backendUtils.generateRequestId(null);
+    String reqId2 = this.backendUtils.generateRequestId(null);
+    assertFalse(reqId1.equals(reqId2));
+  }
+
+  @Test
+  void checkGeneratedWithSameUuidRHDATokenAndDifferentTimeShouldBeDifferentFromEachOther() {
+
+    String randomUUID = UUID.randomUUID().toString();
+    String reqId1 = this.backendUtils.generateRequestId(randomUUID);
+    String reqId2 = this.backendUtils.generateRequestId(randomUUID);
+    assertFalse(reqId1.equals(reqId2));
+  }
+
+  @Test
+  void checkGeneratedWithSameTimeDifferentRHDATokenAndDifferentFromEachOther()
+      throws InterruptedException {
+
+    String randomUUID = UUID.randomUUID().toString();
+    String randomUUID2 = UUID.randomUUID().toString();
+    AtomicReference<String> reqId = new AtomicReference();
+    AtomicReference<String> reqId2 = new AtomicReference();
+    Thread thread = createThread(reqId, randomUUID);
+    Thread thread2 = createThread(reqId2, randomUUID2);
+    thread.start();
+    thread2.start();
+    thread.join();
+    thread2.join();
+    assertFalse(reqId.get().equals(reqId2.get()));
+  }
+
+  @Test
+  void checkGeneratedWithSameTimeSameRHDATokenAndTheyShouldBeEquals() throws InterruptedException {
+
+    String randomUUID = UUID.randomUUID().toString();
+
+    AtomicReference<String> reqId = new AtomicReference();
+    AtomicReference<String> reqId2 = new AtomicReference();
+    Thread thread = createThread(reqId, randomUUID);
+    Thread thread2 = createThread(reqId2, randomUUID);
+    thread.start();
+    thread2.start();
+    thread.join();
+    thread2.join();
+    assertTrue(reqId.get().equals(reqId2.get()));
+  }
+
+  @Test
+  void checkGeneratedWithSameTimeWithNullRHDATokenAndTheyShouldBeDifferent()
+      throws InterruptedException {
+
+    AtomicReference<String> reqId = new AtomicReference();
+    AtomicReference<String> reqId2 = new AtomicReference();
+    Thread thread = createThread(reqId, null);
+    Thread thread2 = createThread(reqId2, null);
+    thread.start();
+    thread2.start();
+    thread.join();
+    thread2.join();
+    assertNotEquals(reqId.get(), reqId2.get());
+  }
+
+  @Test
+  void checkMassGeneratedConcurrentlyWithDifferentRHDATokensAndAllShouldBeDifferentFromEachOther()
+      throws InterruptedException, ExecutionException, TimeoutException {
+    Executor executor = Executors.newCachedThreadPool();
+    List<CompletableFuture<String>> futuresReqIds = new ArrayList<>();
+    Vector<String> results = new Vector<>();
```

Comment:
@ruromero since Set is not thread safe, i used Vector.
Instead, i can use set with putting synchronized block around the insertion to the set collection
```java
    for (int i = 0; i < numberOfRequests; i++) {
      futuresReqIds.add(
          CompletableFuture.supplyAsync(
                  () -> this.backendUtils.generateRequestId(UUID.randomUUID().toString()), executor)
              .whenComplete(
                  (reqId, error) -> {
                    if (results.contains(reqId)) {
                      fail("Found duplicate key: " + reqId);
                    } else {
                      synchronized (this) {
                        results.add(reqId);
                      }
                    }
                  }));
    }
``` 

What do you prefer?